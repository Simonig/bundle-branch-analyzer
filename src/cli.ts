import * as arg from 'arg';
import * as inquirer from 'inquirer';
import { main } from './main';
import { OptionsI } from './types';

async function promptForMissingOptions(options: OptionsI): Promise<OptionsI> {
  const questions = [];

  if (!options.targetBranch) {
    questions.push({
      type: 'input',
      name: 'targetBranch',
      message: 'Please specify the target branch to be compared',
      default: 'master',
    });
  }

  const answers: OptionsI = await inquirer.prompt(questions);
  return {
    ...options,
    targetBranch: options.targetBranch || answers.targetBranch,
    packageManager: 'npm',
  };
}

export function parseArgumentsIntoOptions(rawArgs: string[]): OptionsI {
  const args = arg(
    {
      '--targetBranch': String,
      '--webpackConfig': String,
      '--mode': String,
      '-t': '--targetBranch',
      '-m': '--mode',
    },
    {
      argv: rawArgs.slice(2),
    }
  );
  return {
    targetBranch: args['--targetBranch'] || '',
    webpackConfigScript: args['--webpackConfig'] || '',
    mode: args['--mode'] || null,
  };
}

export async function cli(args: string[]) {
  let options = parseArgumentsIntoOptions(args);
  options = await promptForMissingOptions(options);
  await main(options);
}
