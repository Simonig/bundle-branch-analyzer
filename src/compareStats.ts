import * as path from 'path';
const cwd = process.cwd()
import chalk from 'chalk';
const log = console.log;

function assetsReducer(result: any, value: any) {
    result[value.name] = parseFloat(value.size) / 1024;
    return result;
}

function formatAssets(current: number = 0, target: number = 0, name: string) {
    //console.log({current, target})
    const diff = current - target;
    const percentage = (1 - (current / target)) * 100;

    return {
        name,
        targetSize: target,
        currentSize: current,
        diference: diff,
        pdiff: percentage}
}

function printReport(result: any) {
    log(chalk`Bundle name: {green ${result.name}}
Current branch: {red ${result.currentSize.toFixed(2)} KB}
Target branch: {green ${result.targetSize.toFixed(2)} KB}
Difference: {rgb(255,131,0) ${result.diference.toFixed(2)} KB}
Percentage: {green ${result.pdiff.toFixed(2)}%}
`);
}

export default function compareStats(currentStats: string, targetStats: string) {
    return new Promise((resolve, rejects) => {
        let output = 'bba/'
        let currentStatsPath = path.resolve(cwd, targetStats)
        let targetStatsPath = path.resolve(cwd, currentStats)
        //let outputPath = path.resolve(cwd, output)

        const currentStateAssets = require(currentStatsPath).assets.reduce(assetsReducer, {})
        const targetStateAssets = require(targetStatsPath).assets.reduce(assetsReducer, {})
        const assetsNames = Object.keys({ ...currentStateAssets, ...targetStateAssets });
        //console.log(currentStateAssets);
        let onlyInCurrent = [];
        let onlyInTarget = [];
        let biggerInCurrent = [];
        let smallerInCurrent = [];
        let equal = [];
        let totalDiff = 0;
        let totalPercentage = 0;

        for (let key of assetsNames) {
            const current = currentStateAssets[key];
            const target = targetStateAssets[key];
            const formatted = formatAssets(current, target, key);

            if (!current) {
                onlyInTarget.push(formatted)
            } else if (!target) {
                onlyInCurrent.push(formatted)
            } else if (formatted.diference > 0) {
                biggerInCurrent.push(formatted)
            } else if (formatted.diference < 0) {
                smallerInCurrent.push(formatted)
            } else {
                equal.push(formatted)
            }
            totalDiff += formatted.diference;
            totalPercentage += formatted.pdiff;
        }
        
        log(chalk.black(''))
        log(chalk.blue('only in current'))
        onlyInCurrent.forEach(printReport);

        log(chalk.black('###############################################'))
        log(chalk.black(''))
        log(chalk.blue('only in target'))
        log(chalk.black(''))

        onlyInTarget.forEach(printReport);

        log(chalk.black('###############################################'))

        log(chalk.black(''))
        log(chalk.blue('bigger in current branch'))
        log(chalk.black(''))
        biggerInCurrent.forEach(printReport);

        log(chalk.black('###############################################'))

        log(chalk.black(''))
        log(chalk.blue('smaller in current branch'))
        log(chalk.black(''))

        smallerInCurrent.forEach(printReport);

        log(chalk.black(''))
        log(chalk.black('###############################################'))
        log(chalk.black(''))

        log(chalk.blue(`Total Difference: ${totalDiff.toFixed(2)} KB`))
        log(chalk.blue(`Total Percentage: ${totalPercentage.toFixed(2)}%`))


    })

}