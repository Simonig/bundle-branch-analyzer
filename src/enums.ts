export enum ResultTypes {
  SERVER = 'server',
  STATIC = 'static',
  CONSOLE = 'console',
}
