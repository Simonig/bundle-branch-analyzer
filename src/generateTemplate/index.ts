import * as express from "express";
import { join } from 'path';
import { BundlesReportI } from '../types';



export function runResultExpress(bundleReport: BundlesReportI) {
    const app = express();
    app.set('views', join(__dirname, 'views'));
    app.set('view engine', 'ejs');

    app.get('/', (req: express.Request, res: express.Response) => res.render('result', { bundleReport }));
    const server = app.listen(3333, () => console.log('open http://localhost:3333/ in your browser'));
}