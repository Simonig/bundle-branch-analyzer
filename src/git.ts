import * as inquirer from 'inquirer';
import chalk from 'chalk';
import * as fs from 'fs';
import * as path from 'path';
import * as simplegit from 'simple-git/promise';
import { OptionsI } from './types';

const log = console.log;

const cwd = process.cwd();
const git = simplegit(cwd);

const checkoutOptions = [
  {
    name: `Stash my changes (they can be restored with the command 'git stash pop')`,
    value: 'stash',
  },
  {
    name: `I already fix it, try again`,
    value: 'retry',
  },
  {
    name: `exit`,
    value: 'exit',
  },
];

const errorOnCheckoutQuestion = [
  {
    type: 'list',
    name: 'targetBranch',
    message: 'There was a problem checking out the target branch...',
    choices: checkoutOptions,
  },
];

export async function doCheckout(targetBranch: string) {
  try {
    //log(chalk.blue(`checking out target branch ${options.targetBranch}`));
    await git.checkout(targetBranch);
  } catch (err) {
    const answer = await inquirer.prompt(errorOnCheckoutQuestion);

    switch (answer.targetBranch) {
      case 'stash': {
        log(chalk.blue(`Stashing changes...`));
        await git.stash();
        await doCheckout(targetBranch);
        return;
      }
      case 'retry': {
        await doCheckout(targetBranch);
        return;
      }
      default: {
        process.exit(1);
      }
    }
  }
}

export function getCurrentBranchName(p = process.cwd()): string {
  const gitHeadPath = `${p}/.git/HEAD`;

  return fs.existsSync(p)
    ? fs.existsSync(gitHeadPath)
      ? fs
          .readFileSync(gitHeadPath, 'utf-8')
          .trim()
          .split('/')[2]
      : getCurrentBranchName(path.resolve(p, '..'))
    : '';
}
