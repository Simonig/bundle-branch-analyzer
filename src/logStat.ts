import chalk from 'chalk';
import { BundleStatI, BundlesReportI } from "./types";
const log = console.log;


function printItem(bundleStat: BundleStatI) {
    log(chalk`Bundle name: {green ${bundleStat.name}}
Current branch: {red ${bundleStat.currentSize.toFixed(2)} KB}
Target branch: {green ${bundleStat.targetSize.toFixed(2)} KB}
Difference: {rgb(255,131,0) ${bundleStat.difference.toFixed(2)} KB}
Percentage: {green ${bundleStat.pdiff.toFixed(2)}%}
`);
}


function printReport(bundles: Array<BundleStatI>, title: string) {
    if (bundles.length > 0) {
        log(chalk.black(''))
        log(chalk.blue(title))
        log(chalk.black(''))

        bundles.forEach(printItem);

        log(chalk.black(''))
        log(chalk.black('###############################################'))
        log(chalk.black(''))
    }
}

export default function logStats(bundleReport: BundlesReportI): any {
    const { onlyInCurrent, onlyInTarget, biggerInCurrent, smallerInCurrent, totalDiff, totalPercentage } = bundleReport;
    
    printReport(onlyInCurrent, 'Only in current branch')
    printReport(onlyInTarget, 'Only in target branch')
    printReport(biggerInCurrent, 'Bigger in current branch')
    printReport(smallerInCurrent, 'Smaller in current branch')

    log(chalk.blue(`Total Difference: ${totalDiff.toFixed(2)} KB`))
    log(chalk.blue(`Total Percentage: ${totalPercentage.toFixed(2)}%`))
}