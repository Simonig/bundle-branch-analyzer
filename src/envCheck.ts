import { OptionsI } from './types';
import { getCurrentBranchName } from './git';
import { hasYarn, webpackConfigFound } from './utils';

export function initialEnviromentCheck(options: OptionsI) {
  options.currentBranch = getCurrentBranchName();
  options.packageManager = hasYarn() ? 'yarn' : 'npm';
  if (!webpackConfigFound()) {
    throw new TypeError(`'webpack.config.js' file not found`);
  }
}
