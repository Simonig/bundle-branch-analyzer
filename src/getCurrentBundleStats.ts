import chalk from 'chalk';
import { getCurrentBranchName } from './git';
import { OptionsI } from './types';

const exec = require('await-exec');
const execOpts = { log: false };

export async function getCurrentBundleStats(
  filePrefix: string,
  options: OptionsI
) {
  try {
    const currentBranch = getCurrentBranchName() || filePrefix;
    await exec(`${options.packageManager} install`, execOpts);

    if (options.webpackConfigScript.length > 0) {
      await exec(
        `webpack --config ${options.webpackConfigScript} --profile --json > bba/${currentBranch}-stats.json`,
        execOpts
      );
    } else {
      await exec(
        `webpack --profile --json > bba/${currentBranch}-stats.json`,
        execOpts
      );
    }
  } catch (err) {
    console.log('');
    console.log(
      chalk.red(
        `The following error ocurre while running your webpack configuration:`
      )
    );
    console.error(err);
  }
}
