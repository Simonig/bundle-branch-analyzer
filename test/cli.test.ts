import { parseArgumentsIntoOptions } from '../src/cli';

const BASE_ARGS = [
    '/usr/bin/node',
    '/usr/lib/node_modules/bundle-branch-analyzer.js',
]


describe('cli', () => {
    describe('should parse single arguments', () => {
        it('--targetBranch to targetBranch', () => {
            const parsedArgs = parseArgumentsIntoOptions([...BASE_ARGS, '--targetBranch', 'testArg'])
            expect(parsedArgs.targetBranch).toBe('testArg');
        })

        it('-t to targetBranch', () => {
            const parsedArgs = parseArgumentsIntoOptions([...BASE_ARGS, '-t', 'testArg'])
            expect(parsedArgs.targetBranch).toBe('testArg');
        })

        it('--mode to mode', () => {
            const parsedArgs = parseArgumentsIntoOptions([...BASE_ARGS, '--mode', 'testArg'])
            expect(parsedArgs.mode).toBe('testArg');
        })

        it('-m to mode', () => {
            const parsedArgs = parseArgumentsIntoOptions([...BASE_ARGS, '-m', 'testArg'])
            expect(parsedArgs.mode).toBe('testArg');
        })

        it('--webpackConfig to webpackConfigScript', () => {
            const parsedArgs = parseArgumentsIntoOptions([...BASE_ARGS, '--webpackConfig', 'testArg'])
            expect(parsedArgs.webpackConfigScript).toBe('testArg');
        })

    })

    describe('should parse multiple arguments', () => {
        it('should parse 2 arguments correctly in different order', () => {
            const opts1 = parseArgumentsIntoOptions([...BASE_ARGS, '-t', 'branch', '--mode', 'modeArg'])
            const opts2 = parseArgumentsIntoOptions([...BASE_ARGS, '--mode', 'modeArg', '-t', 'branch',])

            expect(opts1.mode).toBe(opts2.mode)
            expect(opts1.targetBranch).toBe(opts2.targetBranch)
        })

        it('should parse 3 or more arguments correctly in different order', () => {
            const opts1 = parseArgumentsIntoOptions([...BASE_ARGS, '-t', 'branch', '--mode', 'modeArg', '--webpackConfig', 'webpackArg'])
            const opts2 = parseArgumentsIntoOptions([...BASE_ARGS, '--webpackConfig', 'webpackArg', '--mode', 'modeArg', '-t', 'branch',])
            const opts3 = parseArgumentsIntoOptions([...BASE_ARGS, '--mode', 'modeArg', '--webpackConfig', 'webpackArg', '-t', 'branch',])

            expect(opts1.mode).toBe(opts2.mode)
            expect(opts1.mode).toBe(opts3.mode)
            expect(opts2.mode).toBe(opts3.mode)
        })

        it('should keep the options structure', () => {
            const parsedArgs = parseArgumentsIntoOptions([...BASE_ARGS, '-t', 'branch', '--mode', 'modeArg', '--webpackConfig', 'webpackArg'])
            expect(parsedArgs).toMatchSnapshot();
        })


    })
});