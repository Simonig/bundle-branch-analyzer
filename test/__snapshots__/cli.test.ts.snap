// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should parse multiple arguments should set the arguments and keep the default values 1`] = `
Object {
  "mode": "modeArg",
  "outputDir": "bba",
  "targetBranch": "branch",
  "webpackConfigScript": "webpackArg",
}
`;
